packages/intrepid2/unit-test/Projection/test_convergence_HEX.hpp

```diff
@@ -0,0 +1,1154 @@
+        return sin(x*2)*sin(y*2)*cos(z*2)*2+cos(x*y*z*8)*x*y*8;
+      default:
+        return 0;
+      }
+    }
+  };
+
+  struct FunCurl {
+    ValueType
+    KOKKOS_INLINE_FUNCTION
+    operator()(const ValueType& x, const ValueType& y, const ValueType& z, const int comp=0) {
+      ValueType f0 = sin(x*2)*sin(y*2)*sin(z*2)+sin(x*y*z*8);
+      ValueType f1 = cos(x*2)*cos(y*2)*cos(z*2);
+      ValueType f2 = cos(x*y*z*8);
+      //fun = f + a \times x
+      switch (comp) {
+      case 0:
+        return f0;
+      case 1:
+        return f1;
+      case 2:
+        return f2;
+      default:
+        return 0;
+      }
+    }
+  };
+
+  struct CurlFunCurl {
+    ValueType
+    KOKKOS_INLINE_FUNCTION
+    operator()(const ValueType& x, const ValueType& y, const ValueType& z, const int comp=0) {
+      ValueType gf0[3] = {cos(x*2)*sin(y*2)*sin(z*2)*2+cos(x*y*z*8)*y*z*8, sin(x*2)*cos(y*2)*sin(z*2)*2+cos(x*y*z*8)*x*z*8, sin(x*2)*sin(y*2)*cos(z*2)*2+cos(x*y*z*8)*x*y*8};
+      ValueType gf1[3] = {-sin(x*2)*cos(y*2)*cos(z*2)*2, -cos(x*2)*sin(y*2)*cos(z*2)*2, -cos(x*2)*cos(y*2)*sin(z*2)*2};
+      ValueType gf2[3] = {-sin(x*y*z*8)*y*z*8, -sin(x*y*z*8)*x*z*8, -sin(x*y*z*8)*x*y*8};
+      switch (comp) {
+      case 0:
+        return gf2[1] - gf1[2];
+        break;
+      case 1:
+        return gf0[2] - gf2[0];
+        break;
+      case 2:
+        return gf1[0] - gf0[1];
+        break;
+      default:
+        return 0;
+      }
+    }
+  };
+
+
+  struct FunDiv {
+    ValueType
+    KOKKOS_INLINE_FUNCTION
+    operator()(const ValueType& x, const ValueType& y, const ValueType& z, const int comp=0) {
+      ValueType f0 = sin(x*2)*sin(y*2)*sin(z*2)+sin(x*y*z*8);
+      ValueType f1 = cos(x*2)*cos(y*2)*cos(z*2);
+      ValueType f2 = cos(x*y*z*8);
+      //fun = f + a x
+      switch (comp) {
+      case 0:
+        return f0;
+      case 1:
+        return f1;
+      case 2:
+        return f2;
+      default:
+        return 0;
+      }
+    }
+  };
+
+  struct DivFunDiv {
+    ValueType
+    KOKKOS_INLINE_FUNCTION
+    operator()(const ValueType& x, const ValueType& y, const ValueType& z) {
+      ValueType gxf0 = cos(x*2)*sin(y*2)*sin(z*2)*2+cos(x*y*z*8)*y*z*8;
+      ValueType gyf1 = -cos(x*2)*sin(y*2)*cos(z*2)*2;
+      ValueType gzf2 = -sin(x*y*z*8)*x*y*8;
+      return gxf0+gyf1+gzf2;
+    }
+  };
+
+  typedef CellTools<DeviceSpaceType> ct;
+  typedef OrientationTools<DeviceSpaceType> ots;
+  typedef Experimental::ProjectionTools<DeviceSpaceType> pts;
+  typedef RealSpaceTools<DeviceSpaceType> rst;
+  typedef FunctionSpaceTools<DeviceSpaceType> fst;
+
+  constexpr ordinal_type dim = 3;
+  const ordinal_type order = 3;
+  ordinal_type cub_degree = 7;
+
+
+
+  // ************************************ GET INPUTS **************************************
+
+
+  int NX = 2;
+  constexpr int numRefinements = 2;
+
+  ValueType hgradNorm[numRefinements], hcurlNorm[numRefinements], hdivNorm[numRefinements], hvolNorm[numRefinements];
+
```

Comment:
Please add a comment indicating where the values in `hgrad_errors`, etc. come from.

Reply:
Adding this comment:
  // Expected values of the projection errors in H1, Hcurl, Hdiv and L2 norms for HGRAD, HDIV, HCURL and HVOL elements respectively.
  // These values have been computed running the code with numRefinements=4 and the convergence rates are close to the optimal ones.
  // We currently only test two mesh refinements to make the test run faster, so this is used as a regression test rather than
  // a convergence test, but the test can be use for verifying optimal accuracy as well.